### ISSUE TITLE
write_dataset max_open_files does not close oldest file

### ISSUE THREAD

**Issue by xWaita** [apache/arrow#45038]
### Describe the bug, including details regarding any error messages, version, and platform.

According to the docs of pyarrow.dataset.write_dataset, when max_open_files is reached, the least recently used file will be closed (https://arrow.apache.org/docs/python/generated/pyarrow.dataset.write_dataset.html).

However in the C++ code of dataset writer, what actually happens is the largest file is closed (https://github.com/apache/arrow/blob/main/cpp/src/arrow/dataset/dataset_writer.cc#L656). 

For long running dataset writes with many partitions, this means over time the set of open files will trend towards being smaller and smaller in size. Recently opened files will begin to be closed prematurely, while very old open files will hang around and stay in memory. 

It would probably make sense for C++ behaviour to be brought in line with the docs, which should be to close the oldest file rather than the largest file. 

### Component(s)

C++, Python

**Comment by pitrou** [comment]
> It would probably make sense for C++ behaviour to be brought in line with the docs, which should be to close the least recently used file rather than the largest file, with the rationale that the least recently used file is more likely to be finished permanently.

Thanks for the report, and yes, I agree this would be a good idea.

**Comment by pitrou** [comment]
Hmm, the current behavior seems to be deliberate according to the [PR description](https://github.com/apache/arrow/pull/10955):
> `FileSystemDatasetWriteOptions` now has a `max_open_files` setting (ARROW-12321) which prevents opening too many files.  Instead the writer will apply backpressure on the scanner while also closing the open file with the greatest # of rows already written (then resume writing once the file is closed).

@westonpace Do you happen to remember why the option description doesn't match the intended behavior? Is it just a mistake in the docstring?

**Comment by mapleFU** [comment]
This idea LGTM, would you like to implement this?

**Comment by xWaita** [comment]
Thanks for the replies! I can work on it

**Comment by mapleFU** [comment]
> why the option description doesn't match the intended behavior? Is it just a mistake in the docstring?

I guess maybe using same strategy when memory overlimit and fd over limit? Since when we want to reclaim memory, close the largest file ( or k largest file) is usally a good way?

**Comment by pitrou** [comment]
Before working on anything, we should decide which behavior is most desirable. If you want your dataset files to have relatively similar sizes, then it makes sense to close the largest files first.

**Comment by xWaita** [comment]
There is a secondary write_dataset parameter in `max_rows_per_file` that can be used to control how large each file will get. 

For `max_open_files`, it seems the primary intention is to prevent too many open files from building up in memory, in which case it would make sense to choose a file that is most likely to be complete, i.e. least recently used

**Comment by pitrou** [comment]
> With the current implementation of `max_open_files` and writing a dataset with many partitions, rather than seeing file sizes remaining similar, you'd actually see file sizes steadily trend downwards due to small partition files that remain open, while newer partition files compete to stay under the file size of those smaller partitions.

Ah, that may be the case if you have many partitions (more than `max_open_files`) that are very rarely hit?

**Comment by xWaita** [comment]
yeah exactly, and I believe this would be a very common usecase, i.e. if you were to process/write a large file where the partition column is sorted, new partitions would constantly be created while old partitions would never be touched again

**Comment by ryancasburn-KAI** [comment]
As @mapleFU pointed out, it seems like this one parameter (`max_open_files`) is being used for two purposes, keep total file descriptors under a limit and keep memory usage reasonable. 

Perhaps a second parameter should be added: `max_open_rows`. This would control memory usage while `max_open_files` controls fd use.

If `max_open_rows` is reached, the largest file is written (which would drop`max_open_rows` and memory usage the most). If `max_open_files` is reached, the least recently used file is written. This gives people the option to tune the performance as they need. 

`max_open_rows` could be defaulted to `max_rows_per_file * max_open_files`. With this default, the `max_open_rows` would never be reached.

This would help me in work I’m currently doing as I am currently setting a relatively low max_rows_per_file in order to limit memory usage, but this would allow me to set a dataset wide limit allowing individual files to potentially grow larger.
